src/common/cluster/request-api-resources.injectable.ts

```diff
@@ -24,58 +25,49 @@ const requestApiResourcesInjectable = getInjectable({
     const k8sRequest = di.inject(k8SRequestInjectable);
     const logger = di.inject(loggerInjectable);
 
-    return async (cluster) => {
-      const apiLimit = plimit(5);
-      const kubeApiResources: KubeApiResource[] = [];
-      const resourceListGroups: KubeResourceListGroup[] = [];
-
-      try {
-        await Promise.all([
-          (async () => {
-            const { versions } = await k8sRequest(cluster, "/api") as V1APIVersions;
+    const requestApiVersions = async (cluster: Cluster) => (await k8sRequest(cluster, "/api") as V1APIVersions).versions;
+    const requestApisVersions = async (cluster: Cluster) => (await k8sRequest(cluster, "/apis") as V1APIGroupList).groups;
+    const limitingFor = (limit: plimit.Limit) => <Args extends any[], Res>(fn: (...args: Args) => Res) => (...args: Args) => limit(() => fn(...args));
+    const requestKubeApiResourcesFor = (cluster: Cluster) => async ({ group, path }: KubeResourceListGroup): Promise<KubeApiResource[]> => {
+      const { resources } = await k8sRequest(cluster, path) as V1APIResourceList;
 
-            for (const version of versions) {
-              resourceListGroups.push({
-                group: version,
-                path: `/api/${version}`,
-              });
-            }
-          })(),
-          (async () => {
-            const { groups } = await k8sRequest(cluster, "/apis") as V1APIGroupList;
-
-            for (const { preferredVersion, name } of groups) {
-              const { groupVersion } = preferredVersion ?? {};
+      return resources.map(resource => ({
+        apiName: resource.name,
+        kind: resource.kind,
+        group,
+        namespaced: resource.namespaced,
+      }));
+    };
 
-              if (groupVersion) {
-                resourceListGroups.push({
-                  group: name,
-                  path: `/apis/${groupVersion}`,
-                });
-              }
-            }
-          })(),
-        ]);
+    return async (cluster) => {
+      const requestKubeApiResources = requestKubeApiResourcesFor(cluster);
+      const withApiLimit = limitingFor(plimit(5));
 
-        await Promise.all(
-          resourceListGroups.map(({ group, path }) => apiLimit(async () => {
-            const { resources } = await k8sRequest(cluster, path) as V1APIResourceList;
+      try {
+        const resourceListGroups: KubeResourceListGroup[] = [
+          ...(await requestApiVersions(cluster))
+            .map(version => ({
+              group: version,
+              path: `/api/${version}`,
+            })),
+          ...pipeline((await requestApisVersions(cluster)).values())
```

Comment:
Side-note: `pipeline`, as used here, could be more accurately called `chain`, as it implements collection chaining -pattern, and not the pipeline-pattern.

Reply:
ah good point. That is probably a better name.